demos/stackcheck/README.md

```diff
@@ -0,0 +1,45 @@
+
+It is hard to detect illegal use of memory and it may lead to the corruption of the kernel structure. So each time a thread yields, we do checks to ensure the system can work properly.
+
+To illustrate the concept, we need to introduce the following:
+* `check_stack` (in `src/kernel/thread.c`)
+* `choose_next_thread` (in `src/kernel/thread.c`)
+
+## `check_stack`
+
+This function is used to detect if the thread's stack has been overflowed or underflowed. 
+
+In AMT, we place 2 variables `bottom_canary` and `top_canary` on both sides of the stack in the thread structure. These two variables are initialized to `STACK_CANARY`. If their values have been changed, we know that overflow/underflow has occurred.
+
+If the config `KCFG_DESTROY_ON_STACK_ERR` is enabled we invalid the current thread and load the next thread.
+
+If `KCFG_DESTROY_ON_STACK_ERR` is disabled, we just exit the kernel completely. (marking it as an unexpected exit)
+
+## `choose_next_thread`
+
+We use this function to find the next thread to be run. If this function fails to find the next thread, it returns `NULL`.
+
+Before returning the address of the thread, this function checks whether the thread has a right thread ID. If the `tid` differs from the index of the thread in `all_threads`, it prints out the error and exits the system.
+
+## Walkthrough
+
+In the beginning, `setup` thread enables `KCFG_DESTROY_ON_STACK_ERR` to keep the kernel alive when the overflow/underflow occurs.
+
+After setting up, the `setup` creates `watcher` and `overflow` thread and jumps into the `overflow()` function. When we jump into the `overflow()` function, it calls `recurse()` to produce an underflow.
+
+Since we keep calling `recurse()`, the value of the stack pointer will be repeatedly decreased to acquire stack space for these functions. Eventually, the stack will be overwhelmed and variables may be allocated to the address that does not belong to the stack. Thus when we revise the value of these variables, we may alter the structure of the thread. Underflow happens as a result.
+
+After the underflow occurs, the thread is not able to pass `check_stack()` so it will be marked as `INVALID_THREAD` and yield to other threads.
+
+When we enter the `watcher` thread, it tries to join `underflowed` thread. Since the thread has been tagged as `INVALID_THREAD`, the `thread_join()` fails and returns `false`.
+
+After trying to join `overflow`, the thread yields to `underflow`.  Since the `underflow` function is close to the top of the stack (to the bottom of the thread structure), when we write `distance` bytes from the address of `dummy`, we will overwrite the canary. After it yields, the kernel finds its stack is overflowed and then terminates it.
+
+Finally, both `overflow` and `underflow` have been aborted, so the `watcher` fails to join these threads and ends the demo.
```

Comment:
Otherwise this walkthrough looks good.

I did change one of the yields to a yield_to, but this description still fits and we can assume they're looking at the demo source to. So I'd just keep this as is.